### PR TITLE
Fix for TIKA-1787 : NamedEntityParser 

### PR DESCRIPTION
**Pull request by thammegowda**
Added NamedEntityParser that supports loading of different NER implementations at runtime.
The default NER implementation based on OpenNLP is supplied.

~~Another implementation based on StanfordCoreNLP is located [here](https://github.com/thammegowda/tika-ner-corenlp/blob/master/src/test/java/edu/usc/cs/ir/tika/ner/NamedEntityParserTest.java) This is GNU GPL 3, So kept separate.~~ See UPDATE 2 below

@chrismattmann This is not  100% complete, here are few TODOs : 
1. The NER implementing class name needs is to be read from tika config if possible/available. Currently relying on Java Properties. Please suggest me on how to resolve [this todo](https://github.com/thammegowda/tika/commit/e96da2bc28d5eef81d034e39eb05099ed5d38ac1#diff-3a416957f7629e40c8fe5e8fd6c577a3R70)
EDIT : ~~2. Looking for a best way to read parsed text from non text streams within the NamedEntityParser (not sure if a parser can read output of previous parsers like html or pdf). Please suggest me on how to resolve [this todo](https://github.com/thammegowda/tika/commit/e96da2bc28d5eef81d034e39eb05099ed5d38ac1#diff-3a416957f7629e40c8fe5e8fd6c577a3R91)~~ Using secondary parser to get text content

UPDATE : 1. [Added Regex Based NER ](https://github.com/thammegowda/tika/commit/ea7871bd4afae7d18e500ffc285e58afd08f5e86). Though this can recognize much more patterns than names, _(I am using it for recognising weapon names and weapon types )_

UPDATE 2 : [Added Core NLP NER with runtime class binding](https://github.com/thammegowda/tika/commit/084985b3612438e9ca7107fecdffd67757d04d10) this one is still using java binding instead of command invocation, because :
  +  The commandline binding is not portable across environments (or need to maintain those many ports)
  +  setup overead in distribted environment like hadoop.

UPDATE 3 : Chaining support : 
   Now we can chain many NER Implementations (OpenNLP, CoreNLP, RegEx) to the NamedEntityParser.

### CODE REVIEW

File: tika-parsers/src/main/java/org/apache/tika/parser/ner/regex/RegexNERecogniser.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+package org.apache.tika.parser.ner.regex;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.tika.parser.ner.NERecogniser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
Comment (chrismattmann) [inline tika-parsers/src/main/java/org/apache/tika/parser/ner/regex/RegexNERecogniser.java]: please remove star imports

File: tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.tika.parser.ner.opennlp.OpenNLPNERecogniser;
+import org.apache.tika.parser.ner.regex.RegexNERecogniser;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.HashSet;
+
+import static org.junit.Assert.*;
```
Comment (chrismattmann) [inline tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java]: no star imports :)

File: tika-parsers/src/test/java/org/apache/tika/parser/ner/regex/RegexNERecogniserTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.tika.parser.ner.NamedEntityParser;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.junit.Assert.*;
```
Comment (chrismattmann) [inline tika-parsers/src/test/java/org/apache/tika/parser/ner/regex/RegexNERecogniserTest.java]: no star imports

File: tika-parsers/src/test/resources/org/apache/tika/parser/ner/opennlp/ModelGetter.groovy [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    // running from parent maven project, but resources should go to sub-module
+    prefixPath = "tika-parsers/" + prefixPath
+}
+
+def modelFiles = //filePath : url
+        [ (prefixPath + "ner-person.bin"): (urlPrefix + "/en-ner-person.bin"),
+          (prefixPath + "ner-location.bin"): (urlPrefix + "/en-ner-location.bin"),
+          (prefixPath + "ner-organization.bin"): (urlPrefix + "/en-ner-organization.bin"),
+          (prefixPath + "ner-date.bin"): (urlPrefix + "/en-ner-date.bin")/*,
+          (prefixPath + "ner-time.bin"): (urlPrefix + "/en-ner-time.bin"),
```
Comment (chrismattmann) [inline tika-parsers/src/test/resources/org/apache/tika/parser/ner/opennlp/ModelGetter.groovy]: no commented code

File: tika-parsers/src/test/resources/org/apache/tika/parser/ner/tika-config.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chrismattmann) [inline tika-parsers/src/test/resources/org/apache/tika/parser/ner/tika-config.xml]: ALv2 header please

File: tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        //assertTrue(set.contains("California"));
+
+        set.clear();
+        set.addAll(Arrays.asList(md.getValues("NER_ORGANIZATION")));
+        assertTrue(set.contains("University of Southern California"));
+
+        set.clear();
+        set.addAll(Arrays.asList(md.getValues("NER_DATE")));
+        assertTrue(set.contains("1960 - 1975"));
+        //assertTrue(set.contains("1960"));
```
Comment (chrismattmann) [inline tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java]: no commented out code please

File: tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        assertTrue(set.contains(NamedEntityParser.class.getName()));
+
+        set.clear();
+        set.addAll(Arrays.asList(md.getValues("NER_PERSON")));
+        assertTrue(set.contains("John McKay"));
+
+        set.clear();
+        set.addAll(Arrays.asList(md.getValues("NER_LOCATION")));
+        assertTrue(set.contains("Los Angeles"));
+        //assertTrue(set.contains("California"));
```
Comment (chrismattmann) [inline tika-parsers/src/test/java/org/apache/tika/parser/ner/NamedEntityParserTest.java]: no commented out code

File: tika-parsers/src/main/java/org/apache/tika/parser/ner/opennlp/OpenNLPNERecogniser.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.tika.parser.ner.opennlp;
+
+import org.apache.tika.parser.ner.NERecogniser;
+
+import java.util.*;
```
Comment (chrismattmann) [inline tika-parsers/src/main/java/org/apache/tika/parser/ner/opennlp/OpenNLPNERecogniser.java]: no star imports

**Comment by chrismattmann** [comment]
@thammegowda great work! See my comments and please update thank you

**Comment by thammegowda** [comment]
@chrismattmann Thanks for the feedback. Issues Resolved!

File: tika-parsers/src/test/resources/org/apache/tika/parser/ner/opennlp/ModelGetter.groovy [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ */
+def downloadFile(String urlStr, File file) {
+    println "GET : $urlStr -> $file"
+    urlConn = new URL(urlStr).openConnection()
+    contentLength = urlConn.getContentLengthLong()
+
+    file.getParentFile().mkdirs()
+    inStream = urlConn.getInputStream()
+    outStream = new FileOutputStream(file)
+    //IOUtils.copyLarge(inStream, outStream)
```
Comment (chrismattmann) [inline tika-parsers/src/test/resources/org/apache/tika/parser/ner/opennlp/ModelGetter.groovy]: @thammegowda can you remove this line? commented code.

**Comment by chrismattmann** [comment]
one more minor update @thammegowda and this is ready to go!

**Comment by chrismattmann** [comment]
@thammegowda can you also write up a quick tutorial on http://wiki.apache.org/tika/TikaAndNER ? that shows how to install Stanford NER and run this?

**Comment by chrismattmann** [comment]
you will need wiki karma so let me know your username and I'll grant you karma.

**Comment by thammegowda** [comment]
@chrismattmann Sure thing. I might have missed few such comments. I will review one more time. 

Please give me permission to create/edit NER wiki page, my username is "ThammeGowda".

**Comment by chrismattmann** [comment]
done @thammegowda https://wiki.apache.org/tika/ContributorsGroup

**Pull request closed without merging**